larq_compute_engine/tflite/kernels/maxpool.cc

```diff
@@ -0,0 +1,167 @@
+
+  int batches = input->dims->data[0];
+  int height = input->dims->data[1];
+  int width = input->dims->data[2];
+
+  int channels_out = 0;
+  if (input->type == kTfLiteFloat32 || input->type == kTfLiteInt8) {
+    channels_out = GetPackedSize<TBitpacked>(input->dims->data[3]);
+  } else {
+    TF_LITE_ENSURE_EQ(context, input->type, kTfLiteInt32);
+    channels_out = input->dims->data[3];
+  }
+
+  // Allocate temoprary tensor for bitpacked inputs
+  if (input->type == kTfLiteFloat32 || input->type == kTfLiteInt8) {
+    // By default, temporaries is initialized as an array of size 0
+    // The logic here ensures that if we run this code the first time then
+    // its properly re-allocated and when its run a second time (i.e. after a
+    // tensor resize), then we keep the original temporaries array.
+    if (node->temporaries && node->temporaries->size == 0) {
+      TfLiteIntArrayFree(node->temporaries);
+      node->temporaries = nullptr;
+    }
+    if (!node->temporaries) {
+      node->temporaries = TfLiteIntArrayCreate(1);
+      int tensor_id = -1;
+      context->AddTensors(context, 1, &tensor_id);
+      node->temporaries->data[0] = tensor_id;
+    } else {
+      TF_LITE_ENSURE_EQ(context, node->temporaries->size, 1);
+    }
+    TfLiteTensor* packed_input = GetTemporary(context, node, 0);
+
+    TfLiteIntArray* packed_input_size = TfLiteIntArrayCreate(4);
+    packed_input_size->data[0] = batches;
+    packed_input_size->data[1] = height;
+    packed_input_size->data[2] = width;
+    packed_input_size->data[3] = channels_out;
+
+    packed_input->type = kTfLiteInt32;
+    packed_input->allocation_type = kTfLiteArenaRw;
+    TF_LITE_ENSURE_OK(context, context->ResizeTensor(context, packed_input,
+                                                     packed_input_size));
+  }
+
+  // Matching GetWindowedOutputSize in TensorFlow.
+  int out_width, out_height;
+
+  poolparams->padding = ComputePaddingHeightWidth(
+      poolparams->stride_height, poolparams->stride_width, 1, 1, height, width,
+      poolparams->filter_height, poolparams->filter_width,
+      poolparams->padding_type, &out_height, &out_width);
+
+  TfLiteIntArray* output_size = TfLiteIntArrayCreate(4);
+  output_size->data[0] = batches;
+  output_size->data[1] = out_height;
+  output_size->data[2] = out_width;
+  output_size->data[3] = channels_out;
+  return context->ResizeTensor(context, output, output_size);
+}
+
+TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+  BMaxPool2DParams* poolparams =
+      reinterpret_cast<BMaxPool2DParams*>(node->user_data);
+
+  TfLiteTensor* output = GetOutput(context, node, 0);
+  const TfLiteTensor* input = GetInput(context, node, 0);
+
+  const TBitpacked* packed_input_data;
+  RuntimeShape packed_input_shape;
+
+  if (input->type == kTfLiteFloat32) {
+    TfLiteTensor* packed_input = GetTemporary(context, node, 0);
+    ce::core::packbits_tensor<ce::core::BitpackOrder::Canonical>(
```

Comment:
In theory yes, but in this case the packing directly influences the packing of the output, which is 32-bit canonical packing, because thats expected by the next layer.

Reply:
Its a good point that you make actually. We should think about how to do this in the future. Then we will want 'non-canonical' bitpacking for speed, but then the next layer should also know that its input is in 'non-canonical order'.

Reply:
so this ops is only for the case that the next layer is a bcov op expecting a canonical bitpacked input bc currently we don't have non-canonical bitpacking for bitpacked input impl. ?

Reply:
Correct. And thats a good idea :+1: